server/src/main/java/org/opensearch/index/seqno/LocalCheckpointTracker.java

```diff
@@ -131,22 +131,49 @@ public synchronized void markSeqNoAsProcessed(final long seqNo) {
      * @param seqNo the sequence number to mark as persisted
      */
     public synchronized void markSeqNoAsPersisted(final long seqNo) {
-        markSeqNo(seqNo, persistedCheckpoint, persistedSeqNo);
+        markSeqNo(seqNo, persistedCheckpoint, persistedSeqNo, false);
     }
 
-    private void markSeqNo(final long seqNo, final AtomicLong checkPoint, final LongObjectHashMap<CountedBitSet> bitSetMap) {
+    /**
+     * Updates the processed checkpoint to the provided sequence number if segment replication is enabled
+     *
+     * @param seqNo the sequence number to mark as processed
+     */
+    public synchronized void fastForwardProcessedSeqNo(final long seqNo) {
+        markSeqNo(seqNo, processedCheckpoint, null, true);
+    }
+
+    private boolean shouldUpdateSeqNo(final long seqNo, boolean segrep, final AtomicLong checkpoint) {
+        return !((seqNo <= checkpoint.get()) || (segrep && seqNo > persistedCheckpoint.get()));
+    }
+
+    @SuppressForbidden(reason = "Object#notifyAll")
+    private void markSeqNo(
+        final long seqNo,
+        final AtomicLong checkPoint,
+        final LongObjectHashMap<CountedBitSet> bitSetMap,
+        boolean segrep
+    ) {
         assert Thread.holdsLock(this);
         // make sure we track highest seen sequence number
         advanceMaxSeqNo(seqNo);
-        if (seqNo <= checkPoint.get()) {
+        if (shouldUpdateSeqNo(seqNo, segrep, checkPoint) == false) {
             // this is possible during recovery where we might replay an operation that was also replicated
             return;
         }
-        final CountedBitSet bitSet = getBitSetForSeqNo(bitSetMap, seqNo);
-        final int offset = seqNoToBitSetOffset(seqNo);
-        bitSet.set(offset);
-        if (seqNo == checkPoint.get() + 1) {
-            updateCheckpoint(checkPoint, bitSetMap);
+        if (segrep) {
```

Comment:
I don't think we need this segrep specific logic and it can be inside of `fastForwardProcessedSeqNo ` leaving this method as is.  This class shouldn't care about segrep vs non segrep.  

It also looks like `waitForProcessedOpsToComplete` is called from tests only, I'm not sure we have a use case for it with this call path?  If not then we won't need to wait/notify.

Reply:
Removed the wait/notify and separated out the segrep specific logic